lib/msal-node/src/config/ManagedIdentityId.ts

```diff
@@ -0,0 +1,141 @@
+            if (userAssignedClientId || userAssignedResourceId) {
+                createManagedIdentityError(
+                    ManagedIdentityErrorCodes.invalidManagedIdentityIdType
+                );
+            }
+
+            this.id = userAssignedObjectId;
+            this.idType = ManagedIdentityIdType.USER_ASSIGNED_OBJECT_ID;
+            this.isUserAssignedId = true;
+        } else {
+            this.id = DEFAULT_MANAGED_IDENTITY_ID;
+            this.idType = ManagedIdentityIdType.SYSTEM_ASSIGNED;
+            this.isUserAssignedId = false;
+        }
+    }
+
+    /**
+     * Creates an instance of ManagedIdentityId for a system assigned managed identity
+     */
+    public static createSystemAssigned(): ManagedIdentityId {
+        return new ManagedIdentityId();
+    }
+
+    // /**
+    //  * Creates an instance of ManagedIdentityId for a user assigned managed identity from a client id
+    //  * @param clientId Client id of the user assigned managed identity assigned to the azure resource
+    //  * @returns Instance of ManagedIdentityId
+    //  */
+    // public static createUserAssignedClientId(
+    //     clientId: string
+    // ): ManagedIdentityId {
+    //     if (!clientId) {
+    //         throw createManagedIdentityError(
+    //             ManagedIdentityErrorCodes.missingId
+    //         );
+    //     }
+
+    /*
+     *     return new ManagedIdentityId({ userAssignedClientId: clientId });
+     * }
+     */
+
+    // /**
+    //  * Creates an instance of ManagedIdentityId for a user assigned managed identity from a resource id
+    //  * @param resourceId Resource id of the user assigned managed identity assigned to the azure resource
+    //  * @returns Instance of ManagedIdentityId
+    //  */
+    // public static createUserAssignedResourceId(
+    //     resourceId: string
+    // ): ManagedIdentityId {
+    //     if (!resourceId) {
+    //         throw createManagedIdentityError(
+    //             ManagedIdentityErrorCodes.missingId
+    //         );
+    //     }
+
+    /*
+     *     return new ManagedIdentityId({ userAssignedResourceId: resourceId });
+     * }
+     */
+
+    // /**
+    //  * Creates an instance of ManagedIdentityId for a user assigned managed identity from an object id
+    //  * @param objectId Object id of the user assigned managed identity assigned to the azure resource
+    //  * @returns Instance of ManagedIdentityId
+    //  */
+    // public static createUserAssignedObjectId(
+    //     objectId: string
+    // ): ManagedIdentityId {
+    //     if (!objectId) {
+    //         throw createManagedIdentityError(
+    //             ManagedIdentityErrorCodes.missingId
+    //         );
+    //     }
+
+    /*
+     *     return new ManagedIdentityId({ userAssignedObjectId: objectId });
+     * }
+     */
+
+    public get getId(): string {
+        return this.id;
+    }
+
+    public get getIdType(): ManagedIdentityIdType {
```

Comment:
I don't think this should be public API

Reply:
It needs to be public because the type is checked in AppService.ts and IMDS.ts.

Reply:
I don't understand this. We fully control the public API?

Reply:
Could we have `ManagedIdentityId` be only

```json
{
     // empty - means SAMI
}
```

```json
{
     clientId; // UAMI + clientID
}
```

Smth like

```typescript

// This is the public API
export class ManagedIdentityId
{
      public UserAssignedClientId;
      public UserAssignedObjectId;
      public UserAssignedResourceId;
}

// Class that we use in MSAL interally only
class ManagedIdentityIdInternal {

     public readonly type: ManagedIdType 
     constructor(ManagedIdentityId) { 
          type = ... // figure out the type and throw exceptions if incorrect type specified
     }
```     

}

```

Reply:
Even though this class is "exported", it's not available to the developer. In order for an msal-node class or type to be exposed to the developer, it must be exported in lib/msal-node/src/index.ts. In the change-files for this PR, you'll see that only ManagedIdentityApplication and ManagedIdentityRequestParams are exposed (exported).